File: ATG/app/src/main/java/com/kuhakupixel/atg/backend/ACEStatusSubscriber.java
package com.kuhakupixel.atg.backend;

import android.util.Log;

import com.google.gson.Gson;

import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;

import java.io.Closeable;
import java.io.IOException;

public class ACEStatusSubscriber implements Closeable {

    ZContext context = new ZContext();
    ZMQ.Socket subscriber = context.createSocket(SocketType.SUB);
    Gson gson = new Gson();

    public ACEStatusSubscriber(Integer port) throws InterruptedException {
        subscriber.connect(String.format("tcp://localhost:%d", port));
        subscriber.subscribe("");
    }

    public String GetMsg() {
        // trim to remove tailing '0' characters
        return subscriber.recvStr(0).trim();
    }

    public ScanProgressData GetScanProgress() {

        ScanProgressData scanProgressData = new ScanProgressData(0, 0, false);
        try {
            String scanProgress = GetMsg();
            scanProgressData = gson.fromJson(scanProgress, ScanProgressData.class);
        } catch (Exception e) {
            Log.e("ATG", e.getMessage());

        }
        return scanProgressData;
    }


    @Override
    public void close() throws IOException {
        context.close();

    }
}

File: ATG/app/src/main/java/com/kuhakupixel/atg/backend/ACEStatusSubscriber.kt
package com.kuhakupixel.atg.backend

import android.util.Log
import com.google.gson.Gson
import org.zeromq.ZContext
import org.zeromq.ZMQ
import java.io.Closeable

class ACEStatusSubscriber(port: Int) : Closeable {
    var context: ZContext = ZContext()
    var subscriber: ZMQ.Socket = context.createSocket(SocketType.SUB)
    var gson: Gson = Gson()

    init {
        subscriber.connect(String.format("tcp://localhost:%d", port))
        subscriber.subscribe("")
    }

    fun GetMsg(): String {
        // trim to remove tailing '0' characters
        return subscriber.recvStr(0).trim()
    }

    fun GetScanProgress(): ScanProgressData {
        var scanProgressData = ScanProgressData(0, 0, false)
        try {
            val scanProgress = GetMsg()
            scanProgressData = gson.fromJson(scanProgress, ScanProgressData::class.java)
        } catch (e: Exception) {
            Log.e("ATG", e.stackTraceToString())
        }
        return scanProgressData
    }

    override fun close() {
        context.close()
    }
}